File: Advanced/code/problem_19.cpp
// Find the largest rectangular area possible in a given histogram where the largest rectangle can be made of a number of contiguous bars whose heights are given in an array. 
// assume that all bars have the same width and the width is 1 unit.
// eg   histogram = {6, 2, 5, 4, 5, 1, 6}
//  _           _
// | |  _   _  | |
// | | | |_| | | |
// | | | | | | | |
// | |_| | | | | | 
// | | | | | |_| |
// | | | | | | | |
   
// output 4*3=12
   

#include <bits/stdc++.h>

using namespace std;
class Solution {
  public:
    int largestRectangleArea(vector < int > & histo) {
      stack < int > st;
      int maxA = 0;
      int n = histo.size();
      for (int i = 0; i <= n; i++) {
        while (!st.empty() && (i == n || histo[st.top()] >= histo[i])) {
          int height = histo[st.top()];
          st.pop();
          int width;
          if (st.empty())
            width = i;
          else
            width = i - st.top() - 1;
          maxA = max(maxA, width * height);
        }
        st.push(i);
      }
      return maxA;
    }
};
int main() {
  vector < int > histo = {2, 1, 5, 6, 2, 3, 1};
  Solution obj;
  cout << "The largest area in the histogram is " << obj.largestRectangleArea(histo) << endl;
  return 0;
}


// Output: The largest area in the histogram is 10

// Time Complexity: O( N ) + O (N)

// Space Complexity: O(N)

File: Advanced/Python/Problem_19.py
def max_area_histogram(histogram):
	stack = list()

	max_area = 0
	index = 0
	while index < len(histogram):

		if (not stack) or (histogram[stack[-1]] <= histogram[index]):
			stack.append(index)
			index += 1
		else:
			# pop the top
			top_of_stack = stack.pop()

			area = (histogram[top_of_stack] *
					((index - stack[-1] - 1)
					if stack else index))

			max_area = max(max_area, area)

	while stack:

		top_of_stack = stack.pop()

		area = (histogram[top_of_stack] *
				((index - stack[-1] - 1)
				if stack else index))

		max_area = max(max_area, area)


	return max_area


if __name__ == '__main__':
	hist = [6, 2, 5, 4, 5, 1, 6]

	print("Maximum area is",
		max_area_histogram(hist))

